diff --git a/core/adbd/usb_linux_client.c b/core/adbd/usb_linux_client.c
index 0e4d200..34ed069 100644
--- a/core/adbd/usb_linux_client.c
+++ b/core/adbd/usb_linux_client.c
@@ -342,6 +342,9 @@ static void *usb_ffs_open_thread(void *x)
 
         D("[ usb_thread - registering device ]\n");
         register_usb_transport(usb, 0, 0, 1);
+
+        if (access("/sbin/usb_adb_enable.sh", X_OK) == 0)
+            system("/sbin/usb_adb_enable.sh");
     }
 
     // never gets here
